### PR TITLE
tinygobb/builder: mutli go-rountine building and information gathering

### PR DESCRIPTION
**Pull request by leongross**
This PR adds the package tools/tinygobb/builder, which introduces parallel building of single u-root cmdlets with a user-defined compiler (go/tinygo) and build result gathering to compare the information afterward. It should be easy to extend the build information and process it afterward.
After the tinygobb rework done by @archie2x is done, I will integrate this into tinygobb.
Further, this should be integrated into the CI to determine automatically:

* which commands actually built with tinygo
* which errors occur (can be gathered and then publish for each new command)
* compare sized and build time over a longer period of time

@rminnich @archie2x @hugelgupf let me know what you think about this. Also, this is still a work in progress, so I would appreciate your feedback input. 

EDIT: this PR needs `go.122`, maybe it is time now to upgrade u-root to 1.22 or even 1.23?

### CODE REVIEW

**Comment by archie2x** [comment]
I also modified tinygoize to do parallel building and created a .github/workflow to automate building with tinygo.

Here's a run:

https://github.com/archie2x/u-root/actions/runs/10629973266/job/29467948345

I kind of gave up on the whole thing because it didn't seem like the modifications to makebb were going to go in.

File: tools/tinygobb/builder/builder.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+		err := c.Run()
+		t1 := time.Now()
+
+		if err != nil {
+			// the subprocess failed with a non-zero exit code, so create BuildError
+			fmt.Printf("[%d] error building %v:%v", id, job.goPkgPath, errBuf.String())
+			errq <- BuildError{
+				buildJob: job,
+				buildErr: errBuf.String(),
+			}
```
Comment (rminnich) [inline tools/tinygobb/builder/builder.go:221]: suggestion: put a continue here and then you don't need the else.

File: tools/tinygobb/builder/builder.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+	b.logger = logger
+}
+
+// The worker function that will be run by the Builder.
+// The job queue as well as the result and error queue are provided by the builder.
+// TODO: make the channels unidirectional?
+// TODO: how about the mutex syncing for the logger? maybe just leave it out for now
+// func worker(jobq chan BuildJob, resultq chan BuildResult, errq chan BuildError, id uint, logger *log.Logger) {
+func worker(jobq chan BuildJob, resultq chan BuildResult, errq chan BuildError, id uint) {
+	fmt.Printf("[%d] spawned worker %d\n", id, id)
```
Comment (rminnich) [inline tools/tinygobb/builder/builder.go:191]: I'm going to assume you'll remove this logging in the final code?

File: tools/tinygobb/builder/builder.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+// Provide a custom logger for the builder. By default, the logger will be turned off.
+func (b *Builder) SetLogger(logger *log.Logger) {
+	b.logger = logger
+}
+
+// The worker function that will be run by the Builder.
+// The job queue as well as the result and error queue are provided by the builder.
+// TODO: make the channels unidirectional?
```
Comment (rminnich) [inline tools/tinygobb/builder/builder.go:187]: it's easy to do in the declaration of the function, so might as well?

File: tools/tinygobb/builder/builder.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+// TODO: how about the mutex syncing for the logger? maybe just leave it out for now
```
Comment (rminnich) [inline tools/tinygobb/builder/builder.go:188]: why would you need to sync?

File: tools/tinygobb/builder/builder.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+			fmt.Printf("[%d] built %v\n", id, job.goPkgPath)
+			pkgNameToken := strings.Split(job.goPkgPath, "/")
+			binPath := job.goPkgPath + "/" + pkgNameToken[len(pkgNameToken)-1]
+
+			f, err := os.Stat(binPath)
+			if err != nil {
+				fmt.Printf("worker: could not find file %v\n", binPath)
+			}
+
+			resultq <- BuildResult{
```
Comment (rminnich) [inline tools/tinygobb/builder/builder.go:232]: suggstion: you could accumulate all that printing stuff in a bytes.Buffer and return it in the BuildError and BuildResult? that way we don't get noise all over the stdout/stderr.

Also, this error case os.Stat at line 228? should that be on your errq or resultq?

File: tools/tinygobb/builder/builder.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+	// spawn the workers
+	for id := range b.worker {
+		b.logger.Printf("spawning %d workers", b.worker)
+		// go worker(b.jobQueue, b.resultQueue, b.errQueue, id, b.logger)
+		go worker(b.jobQueue, b.resultQueue, b.errQueue, id)
+	}
+
+	// make sure all the workers are spawned and ready.
+	// TODO: make this smarter with a wg or a callback channel
+	time.Sleep(time.Second)
```
Comment (rminnich) [inline tools/tinygobb/builder/builder.go:264]: ouch. Please don't do this Sleep :-) I've started using errgroup and it's pretty nice, consider using it.

File: tools/tinygobb/builder/builder.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+		case err := <-b.errQueue:
+			b.errors = append(b.errors, err)
+			// TODO: why does logger not have Warn or Error?
+			b.logger.Printf("job finished [%d/%d]\n", len(b.errors), nJobs)
+
+		default:
+			// check if all jobs of the jobQueue were sent out. Then stop scheduling
+			// and start only receiving
+			if len(b.jobs) == 0 {
```
Comment (rminnich) [inline tools/tinygobb/builder/builder.go:282]: This seems like an accident waiting to happen. I think you need another way to do whatever it is you're trying to do here :-)

Comment (rminnich) [reply tools/tinygobb/builder/builder.go:282]: I think errgroup will make your life simpler.

**Review by rminnich — COMMENTED** [review]
just a few comments, I think this needs a bit more work :-)

**Comment by leongross** [comment]
We should revive this PR in favor of #3238. I looked at @archie2x's work on the CI part, and it looks very promising.

**Comment by leongross** [comment]
There are some bugs in my multi-go-routine implementation that I currently don't have time to address. I close this PR for now and will create a more streamlined CI utility for regression testing / build status in the next few days. I will work on this again as soon as I have done this properly.

**Pull request closed without merging**